File: packages/caption-schema/src/project.ts
import { z } from "zod";

export const ProjectSourceTypeSchema = z.enum([
  "VIDEO",
  "SUBTITLE",
  "VIDEO_WITH_SUBTITLE",
]);

export type ProjectSourceType = z.infer<typeof ProjectSourceTypeSchema>;

export const ProjectStatusSchema = z.enum([
  "CREATED",
  "PROCESSING",
  "READY",
  "FAILED",
]);

export type ProjectStatus = z.infer<typeof ProjectStatusSchema>;

export const ProjectSchema = z.object({
  id: z.string(),
  name: z.string(),
  sourceType: ProjectSourceTypeSchema,
  status: ProjectStatusSchema.default("CREATED"),
  duration: z.number().nullable().optional(),
  width: z.number().nullable().optional(),
  height: z.number().nullable().optional(),
  language: z.string().nullable().optional(),
  createdAt: z.string().optional(),
  updatedAt: z.string().optional(),
});

export type Project = z.infer<typeof ProjectSchema>;

export const AssetTypeSchema = z.enum(["VIDEO", "AUDIO", "SUBTITLE", "THUMBNAIL"]);

export type AssetType = z.infer<typeof AssetTypeSchema>;

export const ProjectAssetSchema = z.object({
  id: z.string(),
  projectId: z.string(),
  type: AssetTypeSchema,
  originalFilename: z.string(),
  mimeType: z.string(),
  size: z.number(),
  storageKey: z.string(),
  duration: z.number().nullable().optional(),
  width: z.number().nullable().optional(),
  height: z.number().nullable().optional(),
  createdAt: z.string().optional(),
});

export type ProjectAsset = z.infer<typeof ProjectAssetSchema>;

File: apps/api/app/schemas/project.py
from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, Field
from apps.api.app.schemas.caption import CaptionTrackResponse


class AssetResponse(BaseModel):
    id: str
    project_id: str
    type: str
    original_filename: str
    mime_type: str
    size: int
    duration: Optional[float] = None
    width: Optional[int] = None
    height: Optional[int] = None
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ProjectCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    source_type: str = Field(..., pattern="^(VIDEO|SUBTITLE|VIDEO_WITH_SUBTITLE)$")
    language: Optional[str] = "en"


class ProjectResponse(BaseModel):
    id: str
    name: str
    status: str
    source_type: str
    duration: Optional[float] = None
    width: Optional[int] = None
    height: Optional[int] = None
    language: Optional[str] = "en"
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    assets: List[AssetResponse] = []
    caption_tracks: List[CaptionTrackResponse] = []

    class Config:
        from_attributes = True
